Sovereign Chains Cross-Chain Execution (#1033)
multiversx/mx-docs: docs/sovereign/cross-chain-execution.md

> # Creating Bridges

Reviewer: This title is confusing. Creating bridges between what to what should be the answer? At this point in time, a normal user, first time visitor will have no clue to what kind of bridges we are reffering to. Even though we may think that it is obvious (it may be for us) that the topic of this page is the bridge btw MainChain to Sovereign and the other way around, it is not.

When we take a look at the blockchain industry, we observe a segregated ecosystem lacking cohesion, interoperability, teamwork. Many strive to reach the top independently. The vision lead to the Blockchain Revolution, knows as “Web3” — a new era of the internet that is user-centered, emphasizing data ownership and decentralized trust.

Sovereign Chains will dismantle the barriers between isolated blockchain networks by allowing smart contracts to seamlessly interact across different Sovereign Chains and the main MultiversX chain.
This cross-chain interoperability is crucial for fostering an environment where decentralized apps (dApps) can utilize functionalities or assets from across the ecosystem.

## What is Cross-Chain Execution?

> Being able to connect 2 different ecosystems opens up infinite possibilities for functionality. Cross-Chain Execution implies using the functionality of two different chains and combining both of them. In this context we can refer to: 

Reviewer: same thing as above:

A new user will ask himself:
"two chains" - Is it Ethereum and Solana? or what chains are you referring to?

> 1. Sending tokens from one chain to another

Reviewer: Same as above.

Better define from what Chain to what Chain. Probably it will be better to have a note above where you define what MainChain is and what Sovereign Chain is.

2. Executing cross-chain endpoints for chain agnostic use cases

## Cross-Chain Execution within Sovereign Chains

> All the heavy lifting is being done by the *ESDT-Safe* Smart Contract. 

Reviewer: where is this Smart Contract described in detail? if it is not, let's create a page for it. If it is, add a link to that description.

# Introduction

When we take a look at the blockchain industry, we observe a segregated ecosystem lacking cohesion, interoperability, teamwork. Many strive to reach the top independently. The vision lead to the Blockchain Revolution, knows as “Web3” — a new era of the internet that is user-centered, emphasizing data ownership and decentralized trust.

Sovereign Chains will dismantle the barriers between isolated blockchain networks by allowing smart contracts to seamlessly interact across different Sovereign Chains and the main MultiversX chain.
This cross-chain interoperability is crucial for fostering an environment where decentralized apps (dApps) can utilize functionalities or assets from across the ecosystem.

## What is Cross-Chain Execution?

Cross-Chain execution is the ability of smart contracts or decentralized applications on one blockchain to invoke actions on another blockchain. This feature allows for seamless communication and interaction between different blockchain networks, enabling developers to build applications that are chain agnostic.

> ## Cross-Chain Execution within Sovereign Chains

Reviewer 2: I would say we need to add some context about the bridge service because it's needed for my next comments

This feature is enabled by using multiple smart contracts, each one with its unique role and set of functionalities. The current Sovereign Chain suite consists of three main contracts, here is the high-level description for some of the cross chain smart contracts:

#### ESDT-Safe

> The *ESDT-Safe* Smart Contract performs all the heavy lifting. This is the cross-chain execution contract, this execution will be done between either any Sovereign Chain and the MultiversX mainchain or between any Sovereign Chains.

Reviewer 2: - "cross-chain execution contract" - the contract does not execute anything cross-chain. It's responsible for the cross-chain transfers
- "between any Sovereign Chains" - not true

> There are two modules implemented for this contract: [*From Sovereign*](from-sovereign.md), [*To Sovereign*](to-sovereign.md) and two important endpoints: [`execute_operation`](from-sovereign.md#executing-an-operation), [`deposit`](to-sovereign.md#deposit-tokens). Both of them allow the caller to transfer tokens and also execute smart contracts within a single transaction. The only difference is where the sender and receiver of the cross-chain execution are situated. 

Reviewer 2: > Both of them allow the caller to transfer tokens and also execute smart contracts within a single transaction. The only difference is where the sender and receiver of the cross-chain execution are situated.

This is not really accurate. Users only use the `deposit` endpoint to deposit tokens which will be transferred cross-chain. The other endpoint is used only for transfers sov -> main by the bridge service

The scenarios for what endpoint to choose look like this:
1. If the receiver is in a Sovereign Chain, the call should be to the `deposit` endpoint. (_To Sovereign_) 

> 2. If the sender is in a Sovereign Chain, the call should be to the `execute_operation` endpoint. (_From Sovereign_)

Reviewer 2: The call is also on deposit endpoint, then the bridge service will `execute_operation` on main chain

:::note
The naming for those modules has been chosen this way to represent the direction of the execution. In the following sections we will be referring to `FromSovereign` as the execution starts within the Sovereign Chain and `ToSovereign` as the destination of the execution is a Sovereign Chain. 
:::

#### Fee-Market

> Since every Sovereign Chain will have a customizable fee logic, it was paramount that this configuration had to be separated into a different contract. Rules such as: fee token, fee percentages, whitelists are all stored inside this contract.

Reviewer 2: - rules are: fee per transferred token, fee per gas unit and users whitelist to bypass the fee
- you can also add that this contract is present in both chains

#### Header-Verifier

> All the *BLS keys* of the validator will be stored inside this contract. The main role for the Header-Verifier contract is to verify signatures for the operations that are created inside the Sovereign Chain.

Reviewer 2: I would say this with more context, like 
main role is to verify operations - operations are signed by validators - if succesfully verified will be registered and can be executed by ESDT-Safe

:::note

> Each Sovereign Chain will include those smart contracts, the source code can be found at the official [MultiversX Sovereign Chain SCs repository](https://github.com/multiversx/mx-sovereign-sc). 

Reviewer 2: ```suggestion
The source code for the smart contracts can be found at the official [MultiversX Sovereign Chain SCs repository](https://github.com/multiversx/mx-sovereign-sc). 
```

# Introduction

When we take a look at the blockchain industry, we observe a segregated ecosystem lacking cohesion, interoperability, teamwork. Many strive to reach the top independently. The vision lead to the Blockchain Revolution, knows as “Web3” — a new era of the internet that is user-centered, emphasizing data ownership and decentralized trust.

Sovereign Chains will dismantle the barriers between isolated blockchain networks by allowing smart contracts to seamlessly interact across different Sovereign Chains and the main MultiversX chain.
This cross-chain interoperability is crucial for fostering an environment where decentralized apps (dApps) can utilize functionalities or assets from across the ecosystem.

> ## What is Cross-Chain Execution?

Reviewer 3: Create a high level intro here of how sov chains will communicate with main chain. I think the pngs should be added here.

## What is Cross-Chain Execution?

Cross-Chain execution is the ability of a smart contracts or a decentralized applications on one blockchain to invoke actions on another blockchain. This feature allows for seamless communication and interaction between different blockchain networks, enabling developers to build applications that are chain agnostic.

## Cross-Chain Execution within Sovereign Chains

Since a Sovereign Chain is a separate blockchain with a different rule-set from the MultiversX blockchain, there has to be a way of communication between them. The interaction is being done by Smart Contracts, the Sovereign Bridge Service and Nodes. 

![To Sovereign](../../static/sovereign/to-sovereign.png)

When a transaction starts from the MultiversX Mainchain, either from an user or a smart contract, it goes through the `Mvx-ESDT-Safe` smart contract. The Observer nodes pay attention to the events that the deposit transaction emits and then the Sovereign Nodes notarize the state changes inside the Sovereign Chain. This notarization means the end of either the Cross-Chain execution or Cross-Chain transfer.

![From Sovereign](../../static/sovereign/from-sovereign.png)

> If the transaction starts from the Sovereign Chain the call will be to the `Sov-ESDT-Safe` smart contract. The difference is that the output of the transaction is read by the Sovereign Bridge Service. From that point there are two possible outcomes:

Reviewer 3: Don't use an if statement at the start of a paragraph. A reader may choose to read only this paragraph. Don't assume he read all the above info.

When a user or a smart contract want to make a transaction from the Sovereign Chain to MultiversX Mainchain, they must use `Sov-ESDT-Safe` smart contract. The output of the transaction is read by the Sovereign Bridge Service.[...]

1. If the operation is not registered it has to be verified and registered by the `Header-Verifier` smart contract.
2. The Bridge Service calls the `Mvx-ESDT-Safe` smart contract deployed on the MultiversX Mainchain to continue the transaction.

> This is a high-level description of the whole process, the smart contracts that take place in it are far more detailed and have a lot of specific scenarios and behaviours. The current Sovereign Chain suite consists of four main contracts, here is the high-level description for some of the cross chain smart contracts:

Reviewer 3: Link the referred sections. You can refer a section by:
```
[click on this link](#my-multi-word-header)

### My Multi Word Header
```

https://stackoverflow.com/questions/2822089/how-to-link-to-part-of-the-same-document-in-markdown

#### Mvx-ESDT-Safe & Sov-ESDT-Safe

> Those two contracts have the same role: to facilitate a cross-chain execution depending on what side the process starts. There will be an in-depth description of each smart contract in the upcoming modules.

Reviewer 3: "The two contracts [...]"

#### Fee-Market

> Since every Sovereign Chain will have a customizable fee logic, it was paramount that this configuration had to be separated into a different contract. Rules such as: fee per transferred token, fee per gas unit and users whitelist to bypass the fee are set here. This contract is also present in the MultiversX mainchain and in any Sovereign Chain.

Reviewer 3: Put the rules as a list.

# Introduction

When we take a look at the blockchain industry, we observe a segregated ecosystem lacking cohesion, interoperability and teamwork. The vision lead to the Blockchain Revolution, knows as “Web3” — a new era of the internet that is user-centered, emphasizing data ownership and decentralized trust.

Sovereign Chains will dismantle the barriers between isolated blockchain networks by allowing smart contracts to seamlessly interact across different Sovereign Chains and the main MultiversX chain.
This cross-chain interoperability is crucial for fostering an environment where decentralized apps (dApps) can utilize functionalities or assets from across the ecosystem.

## What is Cross-Chain Execution?

Cross-Chain execution is the ability of a smart contracts or a decentralized applications on one blockchain to invoke actions on another blockchain. This feature allows for smart contract execution or transfer of funds from one chain to another, enabling developers to build applications that are chain agnostic.

## Cross-Chain Execution within Sovereign Chains

> Since a Sovereign Chain is a separate blockchain with a different rule-set from the MultiversX blockchain, there has to be a way of communication between them. The interaction is being done by Smart Contracts, the Sovereign Bridge Service and Nodes. 

Reviewer 3: MultiversX blockchain (will also be named MultiversX Mainchain in the further sections)

![To Sovereign](../../static/sovereign/to-sovereign.png)

> When a transaction starts from the MultiversX Mainchain, either from a wallet or a smart contract, it goes through the `Mvx-ESDT-Safe` smart contract. The Observer nodes pay attention to the events that the deposit transaction emits and then the Sovereign Nodes notarize the state changes inside the Sovereign Chain. This notarization means the end of either the Cross-Chain execution or Cross-Chain transfer.

Reviewer 3: I would change `pay attention` to `listen` or `monitor`.

![From Sovereign](../../static/sovereign/from-sovereign.png)

> When the transaction sender is from the Sovereign Chain, the `Sov-ESDT-Safe` smart contract must be used. The output of the transaction is read by the Sovereign Bridge Service. From that point there are two possible outcomes:

Reviewer 3: I would change `When the transaction sender is from the Sovereign Chain` -> `When a transaction is generated from the Sovereign Chain`. 
The word `sender` here creates ambiguity.

2. If the operation is registered, the Sovereign Bridge Service calls the `Mvx-ESDT-Safe` smart contract deployed on the MultiversX Mainchain to continue the transaction.

This is a high-level description of the whole process, the smart contracts that take place in it are far more detailed and have a lot of specific scenarios and behaviours. The current Sovereign Chain suite consists of four main contracts, here is the high-level description for some of the cross chain smart contracts:

## Mvx-ESDT-Safe & Sov-ESDT-Safe
The two contracts have the same role: to facilitate a cross-chain execution depending on what side the process starts. The reason for the prefix of `Sov` and `Mvx` is to show where the smart contract is deployed, `Sov` means that the contract is deployed on a Sovereign Chain and `Mvx` that is deployed on the MultiversX Mainchain. There will be an in-depth description of each smart contract in the upcoming modules ([`Mvx-ESDT-Safe`](mvx-esdt-safe.md) and [`Sov-ESDT-Safe`](sov-esdt-safe.md)). The description will consist of flows for the cross-chain interactions, important modules and endpoints. 

## Fee-Market

> Since every Sovereign Chain will have a customizable fee logic, it was paramount that this configuration had to be separated into a different contract. There rules set inside this contract are: 

Reviewer 3: `There` I think it should be `The`.

* fee per transferred token 
* fee per gas unit 
* users whitelist to bypass the fee

> This contract is also present in the MultiversX mainchain and in any Sovereign Chain.

Reviewer 3: `MultiversX mainchain` - > `MultiversX Mainchain`. Be consistent with the naming across all sections

## Cross-Chain Execution within Sovereign Chains

Since a Sovereign Chain is a separate blockchain with a different rule-set from the MultiversX blockchain, there has to be a way of communication between them. The interaction is being done by Smart Contracts, the Sovereign Bridge Service and Nodes. 

> The MultiversX blockchain will be referred as the _MultiversX Mainchain_ in the further sections.

![To Sovereign](../../static/sovereign/to-sovereign.png)

When a transaction starts from the MultiversX Mainchain, either from a wallet or a smart contract, it goes through the `Mvx-ESDT-Safe` smart contract. The Observer nodes monitor the events that the deposit transaction emits and then the Sovereign Nodes notarize the state changes inside the Sovereign Chain. This notarization means the end of either the Cross-Chain execution or Cross-Chain transfer.

![From Sovereign](../../static/sovereign/from-sovereign.png)

> When the transaction is generated from the Sovereign Chain, the `Sov-ESDT-Safe` smart contract must be used. The output of the transaction is read by the Sovereign Bridge Service. From that point there are two possible outcomes:

Reviewer 2: This part is not really correct.
- The SC will generate some events
- Then sovereign nodes, at the end of the round, will read all the outgoing operations (plus all the unconfirmed operations), then these are signed by all the validators, then send to the bridge service, which will make the transaction on main chain
